### PR TITLE
Fix issue 22920: Improve errors when missing import paths

### PR DESCRIPTION
**Pull request by ljmf00**
This patch introduces errorSupplemental() to the import paths error to display
    extra information about them. These error messages are out of sync with `-J`
    flag, which have similar purposes. This patch make `-I` behave similarly to
    `-J` flag in terms of errors, improving consistency among other error messages
    and add some alignment to the main error.
    
    It also improves the message for missing `object.d` file and give better hints
    about configured import paths.
    
    This change is particularly crucial for DMD as a library, since error
    interfaces writes to a diagnostic handler rather than directly to `stderr`.
    
    Signed-off-by: Luís Ferreira <contact@lsferreira.net>

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
How does this improve the error message - the import path is unrelated to the location of the import?
Please include such information in the commit title / message (the same applies to your other PR's btw) - see e..g [rule 7 in this article](https://cbea.ms/git-commit/)

**Comment by ljmf00** [comment]
> How does this improve the error message - the import path is unrelated to the location of the import? Please include such information in the commit title / message (the same applies to your other PR's btw) - see e..g [rule 7 in this article](https://cbea.ms/git-commit/)

Sometimes I forgot to add a proper description but this one in specific is still in draft. I strive to do it tho. Thank for the heads up :)

File: src/dmd/dmodule.d [hunk 1]
```diff
@@ -601,10 +601,21 @@ extern (C++) final class Module : Package
...
+                errorSupplemental(loc, "Perhaps add an include path using `-I` or into the config file (`%.*s`)",
+                        cast(int)dmdConfFile.length,
+                        dmdConfFile.ptr);
+            }
+            else
+            {
+                errorSupplemental(loc, "There is no config file. Perhaps add one or add an include path using `-I`.");
+            }
+            if (global.path)
+                errorSupplemental(loc, "Currently, the compiler found the following include paths:");
```
Comment (ibuclaw) [inline src/dmd/dmodule.d:618]: The compiler doesn't "find" include paths, suggests that the wording is misleading.

File: src/dmd/dmodule.d [hunk 2]
```diff
@@ -623,26 +634,32 @@ extern (C++) final class Module : Package
...
-             */
-            if (global.path)
+            foreach (idx, path; *global.path)
             {
-                foreach (i, p; *global.path)
-                    fprintf(stderr, "import path[%llu] = %s\n", cast(ulong)i, p);
+                const attr = FileName.exists(path);
+                const(char)* err = attr == 2 ? "" :
+                    (attr == 1 ? " (not a directory)" : " (path not found)");
+                .errorSupplemental(loc, "[%llu]: `%s`%s", cast(ulong)idx, path, err);
```
Comment (ibuclaw) [inline src/dmd/dmodule.d:654]: errorSupplemental(loc) would be the wrong interface for getting this information out. Try compiling any of the examples with `-verrors=context` for a real comparison between the old and new diagnostic messages.

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/dmd/dmodule.d (hunk 2 above)
Comment (ljmf00) [reply src/dmd/dmodule.d:654]: What do you suggest?

File: src/dmd/dmodule.d (hunk 1 above)
Comment (ljmf00) [reply src/dmd/dmodule.d]: I changed to `search on`. If you know better wording, please tell

File: src/dmd/dmodule.d (hunk 2 above)
Comment (ljmf00) [reply src/dmd/dmodule.d:654]: `errorSupplemental` should remember the previous location and avoid repeating printing the context location when `-verrors=context`. This happens for other `errorSupplemental` errors. I would consider this an improvement to `errorSupplemental` with `-verrors=context` in a subsequent patch.

Comment (ibuclaw) [reply src/dmd/dmodule.d:654]: > What do you suggest?

message() is good for verbatim messages, but dmd wants to be different and sends them to stdout instead of stderr.

If it can't be expressed in one line, I'd probably reference a way to dump search paths of the compiler.

Comment (ljmf00) [reply src/dmd/dmodule.d:654]: > > What do you suggest?
> 
> message() is good for verbatim messages, but dmd wants to be different and sends them to stdout instead of stderr.

Well, message() doesn't write to `DiagnosticHandler`. My initial motivation was that import paths were being written directly with `fprintf` and I wanted them to be written to `DiagnosticHandler` so I can use DMD as a library to create a linter. I can make a PR to fix that.

> If it can't be expressed in one line, I'd probably reference a way to dump search paths of the compiler.

What do you mean? A compiler flag to dump the search paths? I think that is not helpful from the user's perspective. If I'm compiling something the build system will probably specify the include paths via command line, and would probably be easier to get them from the build tool. I think the verbosity flag can be used to limit the number of include paths.

Comment (ibuclaw) [reply src/dmd/dmodule.d:654]: > > > What do you suggest?
> > 
> > message() is good for verbatim messages, but dmd wants to be different and sends them to stdout instead of stderr.
> 
> Well, message() doesn't write to `DiagnosticHandler`. My initial motivation was that import paths were being written directly with `fprintf` and I wanted them to be written to `DiagnosticHandler` so I can use DMD as a library to create a linter. I can make a PR to fix that.
> 

Maybe, dmd can do what it sees fit here. All dmd.errors entrypoints go through the gcc diagnostic engine in gdc.

> > If it can't be expressed in one line, I'd probably reference a way to dump search paths of the compiler.
> 
> What do you mean? A compiler flag to dump the search paths? I think that is not helpful from the user's perspective. If I'm compiling something the build system will probably specify the include paths via command line, and would probably be easier to get them from the build tool. I think the verbosity flag can be used to limit the number of include paths.

An example of a bad diagnostic error would be:
```
var.error("conflicts with `%s` at %s", other.toPrettyChars(), other.locToChars());
```
A better example that makes good use of errorSupplemental would be:
```
var.error("conflicts with `%s`", other.toPrettyChars());
errorSupplemental(other.loc, "previous definition here");
```

All uses of error and errorSupplemental should take into account that caret/source line printing is default enabled, so you should avoid calling it too repetitively, include newlines in diagnostic messages, or attempt to format the message in a peculiar way such as vertically aligning them, as it will be rendered horrible and broken up by the two source+underscore lines between each message.

We already have an error to the effect of `%d instantiations, recompile with -v to show more` for instantiation traces deeper than ~6.  There could be a similar one here too - in fact, I'm not sure whether `-v` dumps all search paths, it probably should given everything else it does.

Comment (ljmf00) [reply src/dmd/dmodule.d:654]: > All uses of error and errorSupplemental should take into account that caret/source line printing is default enabled, so you should avoid calling it too repetitively, include newlines in diagnostic messages, or attempt to format the message in a peculiar way such as vertically aligning them, as it will be rendered horrible and broken up by the two source+underscore lines between each message.

Right, I understand the purpose now. Although I think we misuse `errorSupplemental` in some places then. We could have an alternative that does the error alignment but doesn't show the carret, just to add additional information.

> We already have an error to the effect of `%d instantiations, recompile with -v to show more` for instantiation traces deeper than ~6. There could be a similar one here too - in fact, I'm not sure whether `-v` dumps all search paths, it probably should given everything else it does.

We can still show the import paths, but only show like 3 or at maximum 6 or something we define. I think not showing them is not practical at all.

File: src/dmd/dmodule.d [hunk 3]
```diff
@@ -601,10 +601,21 @@ extern (C++) final class Module : Package
     {
         if (FileName.equals(srcfile.toString(), "object.d"))
         {
-            .error(loc, "cannot find source code for runtime library file 'object.d'");
-            errorSupplemental(loc, "dmd might not be correctly installed. Run 'dmd -man' for installation instructions.");
-            const dmdConfFile = global.inifilename.length ? FileName.canonicalName(global.inifilename) : "not found";
-            errorSupplemental(loc, "config file: %.*s", cast(int)dmdConfFile.length, dmdConfFile.ptr);
+            error("cannot find source code for runtime library file `object.d`.");
+            /* errorSupplemental("DMD might not be correctly installed. Run 'dmd -man' for installation instructions."); */
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply src/dmd/dmodule.d:605]: Maybe you can just delete this instead of commenting.

File: src/dmd/dmodule.d [hunk 4]
```diff
@@ -601,10 +601,21 @@ extern (C++) final class Module : Package
...
+                errorSupplemental(loc, "Perhaps add an include path using `-I` or into the config file (`%.*s`)",
+                        cast(int)dmdConfFile.length,
+                        dmdConfFile.ptr);
+            }
+            else
+            {
+                errorSupplemental(loc, "There is no config file. Perhaps add one or add an include path using `-I`.");
+            }
+            if (global.path)
+                errorSupplemental(loc, "Currently, the compiler search on the following include paths:");
```
Comment (RazvanN7) [inline src/dmd/dmodule.d:618]: I would rephrase this to: "Current include paths:"

**Comment by RazvanN7** [comment]
@ljmf00 Any updates on this PR?
